### PR TITLE
[SPARK-46270][SQL][CORE][SS] Use java16 instanceof expressions to replace the java8 instanceof statement.

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
This PR uses java14 `instanceof` to replace the java8 `instanceof`.
For example:
```
if (obj instanceof String) {
    String s = (String) obj;    // grr...
    ...
}
```
We can change it to
```
if (obj instanceof String s) {
    // Let pattern matching do the work!
    ...
}
```

### Why are the changes needed?
Using [`[JEP 394: Pattern Matching for instanceof]` ](https://openjdk.org/jeps/394)can bring the following benefits:

1. **More concise syntax**: Pattern matching allows the desired "shape" of an object to be expressed concisely (the pattern), and for various statements and expressions to test that "shape" against their input (the matching).

2. **Safer**: The motto is: "A pattern variable is in scope where it has definitely matched". This allows for the safe reuse of pattern variables and is both intuitive and familiar, since Java developers are already used to flow sensitive analyses.

3. **Avoid explicit casts**: The use of pattern matching in instanceof should significantly reduce the overall number of explicit casts in Java programs. Type test patterns are particularly useful when writing equality methods.

### Does this PR introduce _any_ user-facing change?
'No'.

### How was this patch tested?
GA

### Was this patch authored or co-authored using generative AI tooling?
'No'.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for this change, @beliefer .

**Comment by dongjoon-hyun** [comment]
BTW, @beliefer , JEP 394 is Java 16 officially. Could you revise the following PR description?
> This PR uses java14 instanceof to replace the java8 instanceof.

Screenshot 2023-12-05 at 8 05 45 AM

**Comment by dongjoon-hyun** [comment]
Ya, Java 14 and 15 has different JEP IDs.
- Java 14: [JEP 305: Pattern Matching for instanceof (Preview)](https://openjdk.org/jeps/305)
- Java 15: [JEP 375: Pattern Matching for instanceof (Second Preview)](https://openjdk.org/jeps/375)

**Comment by beliefer** [comment]
The GA failure is unrelated.

**Comment by beliefer** [comment]
cc @LuciferYang

**Review by LuciferYang — COMMENTED** [review]
others is fine to me, only one comments

File: sql/hive/src/test/java/org/apache/spark/sql/hive/test/Complex.java [hunk 1]
```diff
@@ -547,8 +547,8 @@ public boolean equals(Object that) {
     if (that == null) {
       return false;
     }
-    if (that instanceof Complex) {
-      return this.equals((Complex)that);
+    if (that instanceof Complex complex) {
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/hive/src/test/java/org/apache/spark/sql/hive/test/Complex.java:550]: OK

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you, @beliefer and @LuciferYang .

**Comment by beliefer** [comment]
@dongjoon-hyun @LuciferYang Thank you.

**Comment by cloud-fan** [comment]
Hi @beliefer , thanks for your time to clean up the Spark codebase! However, I have a bit of concern about this kind of invasive changes. The new API is good and we should definitely use them when writing code. But do we really need to rewrite the old code? Java has a good reputation for backward compatibility, and these old code won't stop us from upgrading to new Java versions. It does bring troubles when a PR touches these files and needs to backport. I think it's not worth our time to rewrite the old code, unless it blocks us from upgrading Scala/Java.

I think a simple principle is to not touch the old code if they are working well and not deprecated. Is the old `instanceof` syntax deprecated in new Java versions?

**Pull request closed without merging**